src/interface/pipe.rs

```diff
@@ -97,7 +98,10 @@ impl EmulatedPipe {
         }
 
         while bytes_written < length {
-            let bytes_to_write = min(length, bytes_written as usize + write_end.remaining());
+            let remaining = write_end.remaining();
+            // we write if the pipe is empty, otherwise we try to limit writes to 4096 bytes (unless whats leftover of this write is < 4096)
+            if remaining != self.size  && (length - bytes_written) > PAGE_SIZE && remaining < PAGE_SIZE { continue };
```

Comment:
So this was a busy loop before and is still a busy loop now?
Is there no possibility of a TOCTTOU attack here?
Not entirely sure about the semantics but if there's no mutual exclusion that seems like it would be possible?

Reply:
So because of the mutexes around write/read we can assume that theres only one active writer until it completes and/or one active reader.

remaining()/len() both give the amount we can work with in the pipe, either an amount of space to write to for the writer, or an amount to read from for the reader. So we check remaining - then a reader reads some - then we write, we still have at least that amount to write to (but would be more now). This works in the opposite direction as well.

Basically there isn't a TOCTTOU for moving the head/available length for reads, or tail/available space for writes.

Reply:
Ok that makes sense to me! Looks good.